src/hook.c

```diff
@@ -0,0 +1,153 @@
+/*
+ * Copyright (C) the libgit2 contributors. All rights reserved.
+ *
+ * This file is part of libgit2, distributed under the GNU GPL v2 with
+ * a Linking Exception. For full terms see the included COPYING file.
+ */
+
+#include "common.h"
+#include "posix.h"
+#include "repository.h"
+
+#include "git2/hook.h"
+
+#define GIT_HOOKS_DIR "hooks/"
+
+const char * const githooks[] = {
+	"applypatch-msg",
+	"pre-applypatch",
+	"post-applypatch",
+	"pre-commit",
+	"prepare-commit-msg",
+	"commit-msg",
+	"post-commit",
+	"pre-rebase",
+	"post-checkout",
+	"post-merge",
+	"pre-push",
+	"pre-receive",
+	"update",
+	"post-receive",
+	"post-update",
+	"push-to-checkout",
+	"pre-auto-gc",
+	"post-rewrite",
+};
+
+static int hook_dir(git_buf *out_dir, git_repository *repo)
+{
+	int err;
+	git_buf tmp_path = GIT_BUF_INIT;
+	git_buf cfg_path = GIT_BUF_INIT;
+	git_config *cfg;
+
+	assert(out_dir && repo);
+
+	/* We need to check for an override in the repo config */
+	err = git_repository_config__weakptr(&cfg, repo);
+	if (err != 0)
+		return err;
+
+	err = git_config_get_path(&cfg_path, cfg, "core.hooksPath");
+	if (err == GIT_ENOTFOUND) {
+		giterr_clear();
+		git_buf_puts(&tmp_path, git_repository_commondir(repo));
+		git_buf_joinpath(&tmp_path, tmp_path.ptr, GIT_HOOKS_DIR);
+	} else if (err == GIT_OK) {
+		git_buf_joinpath(&tmp_path, git_repository_commondir(repo), cfg_path.ptr);
+		git_path_resolve_relative(&tmp_path, 0);
+		git_buf_free(&cfg_path);
+	} else {
+		return -1;
+	}
+
+	*out_dir = tmp_path;
+	return 0;
+}
+
+int git_hook_dir(char **out_dir, git_repository *repo) {
+	git_buf hook_dir_buf = GIT_BUF_INIT;
+	int err;
+
+	assert(out_dir != NULL);
+
+	err = hook_dir(&hook_dir_buf, repo);
+	if (err != 0) {
+		*out_dir = NULL;
+		return -1;
+	}
+
+	*out_dir = git_buf_detach(&hook_dir_buf);
+
+	return 0;
+}
+
+static int build_hook_path(char **out_path, git_repository *repo, const char *hook_name)
+{
+	int err;
+	git_buf hook_path = GIT_BUF_INIT;
+
+	assert(hook_name);
+
+	err = hook_dir(&hook_path, repo);
+	if (err != 0)
+		return -1;
+
+	err = git_buf_joinpath(&hook_path, hook_path.ptr, hook_name);
+	if (err != 0)
+		return -1;
+
+	*out_path = git_buf_detach(&hook_path);
+
+	return 0;
+}
+
+static int check_hook_path(const char *hook_path)
+{
+	int err;
+	struct stat hook_stat;
+
+	assert(hook_path);
+
+	/* Skip missing hooks */
+	err = p_stat(hook_path, &hook_stat);
+	if (err)
```

Comment:
I'm a bit undecided here. I feel like we should check specifically for `ENOTFOUND` here and handle other errors as "real" errors. What do you think?

Reply:
Agreed. I'll do a mapping of `ENOENT` to `GIT_ENOTFOUND` and the rest to a generic `GITERR_OS` class.